aws_doc_sdk_examples_tools/doc_gen.py

```diff
@@ -36,58 +41,126 @@ class DocGen:
     examples: List[Example] = field(default_factory=list)
     cross_blocks: Set[str] = field(default_factory=set)
 
-    def collect_snippets(self, snippets_root: Optional[Path]):
+    def collect_snippets(
+        self, snippets_root: Optional[Path] = None, prefix: Optional[str] = None
+    ):
+        if prefix is not None:
+            prefix = f"{prefix}_"
+        if prefix is None:
+            prefix = ""
         if snippets_root is None:
-            snippets_root = self.root.parent.parent
-        snippets, errs = collect_snippets(snippets_root)
+            snippets_root = self.root
+        snippets, errs = collect_snippets(snippets_root, prefix)
         self.snippets = snippets
         self.errors.extend(errs)
 
+    def languages(self) -> Set[str]:
+        languages: Set[str] = set()
+        for sdk_name, sdk in self.sdks.items():
+            for version in sdk.versions:
+                languages.add(f"{sdk_name}:{version.version}")
+        return languages
+
+    def merge(self, other: "DocGen") -> MetadataErrors:
+        """Merge fiends from other into self, prioritizing self fields"""
+        warnings = MetadataErrors()
+        for name, sdk in other.sdks.items():
+            if name not in self.sdks:
+                self.sdks[name] = sdk
+            else:
+                warnings.append(
+                    DocGenMergeWarning(
+                        file=str(other.root), id=f"conflict in sdk {name}"
+                    )
+                )
+        for name, service in other.services.items():
+            if name not in self.services:
+                self.services[name] = service
+                warnings.append(
+                    DocGenMergeWarning(
+                        file=str(other.root), id=f"conflict in service {name}"
+                    )
+                )
+        for name, snippet in other.snippets.items():
+            if name not in self.snippets:
+                self.snippets[name] = snippet
+                warnings.append(
+                    DocGenMergeWarning(
+                        file=str(other.root), id=f"conflict in snippet {name}"
+                    )
+                )
+
+        self.snippet_files.update(other.snippet_files)
+        self.cross_blocks.update(other.cross_blocks)
+        self.examples += other.examples
```

Comment:
I think this won't be right, because examples each have a unique key ID, so putting them all in a list means you might have duplicates, and what happens if you have a duplicate later in the list?
The current method is to only merge the language subsection if it doesn't exist, and otherwise reject a duplicate example (and don't mere any other fields like title, etc., though it's not an error for them to exist).

Reply:
I don't like merging at the language level because I can see a service team coming along (or a guide writer) needing to add a specific example for a specific language in the bull pen repo.

Reply:
Though we could also argue that the key is (file, id)